File: py/fb_encoder.py

import flatbuffers
import sys

import Person
import Group
import Client
import ClientList
from ClientUnion import ClientUnion 


builder = flatbuffers.Builder(1024)

# creating person 'Ram'
male_str = builder.CreateString("Male")
female_str = builder.CreateString("Female")

person1_name = builder.CreateString("Ram")

Person.Start(builder)
Person.AddName(builder,person1_name)
Person.AddAge(builder,21)
Person.AddWeight(builder,76.5)
Person.AddGender(builder,male_str)
person1_ram = Person.End(builder)

# Creating group 'FightClub'
person2_name = builder.CreateString("Shyam")
person3_name = builder.CreateString("Raghuveer")

Group.StartNamesVector(builder,3)
builder.PrependUOffsetTRelative(person1_name)
builder.PrependUOffsetTRelative(person2_name)
builder.PrependUOffsetTRelative(person3_name)
person_names_in_group = builder.EndVector()

group1_name = builder.CreateString("FightClub")
Group.Start(builder)
Group.AddNames(builder,person_names_in_group)
Group.AddAverageAge(builder,24.5)
Group.AddAverageWeight(builder,66.0)
Group.AddGroupname(builder,group1_name)
group1 = Group.End(builder)

# Add Person as Client
Client.Start(builder)
Client.AddClientType(builder,ClientUnion.Person)
Client.AddClient(builder,person1_ram)
client1 = Client.End(builder)

# Add Group as Client
Client.Start(builder)
Client.AddClientType(builder,ClientUnion.Group)
Client.AddClient(builder,group1)
client2 = Client.End(builder)

# Create Vector of Clients
ClientList.StartClientsVector(builder,2)
builder.PrependUOffsetTRelative(client2)
builder.PrependUOffsetTRelative(client1)
client_list = builder.EndVector()

ClientList.Start(builder)
ClientList.AddClients(builder,client_list)
clients = ClientList.End(builder)

builder.Finish(clients)

buf = builder.Output()

# Write as bin file
open(sys.argv[1],'wb').write(buf);

File: cpp/fb_encoder.cpp

#include "schema_generated.h"
#include <fstream>
#include <iostream>

int main(int argc, const char *argv[]) {
    // Build up a serialized buffer algorithmically:
    flatbuffers::FlatBufferBuilder builder;

    // creating person 'Ram'
    auto male_str = builder.CreateString("Male");
    auto female_str = builder.CreateString("Female");

    auto person1_name = builder.CreateString("Ram");

    auto person1_ram = CreatePerson(builder,person1_name,21,76.5,male_str);
    
    // Creating group 'FightClub'
    auto person2_name = builder.CreateString("Shyam");
    auto person3_name = builder.CreateString("Raghuveer");

    std::vector<flatbuffers::Offset<flatbuffers::String>> names_vector; 
    names_vector.push_back(person1_name);
    names_vector.push_back(person2_name);
    names_vector.push_back(person3_name);
    auto person_names_in_group = builder.CreateVector(names_vector);

    auto group1_name = builder.CreateString("FightClub");

    auto group1 = CreateGroup(builder,group1_name,24.5, 66.0,person_names_in_group);
    
    // Add Person as Client
    auto client1 = CreateClient(builder,ClientUnion_Person,person1_ram.Union());


    // Add Group as Client
    auto client2 = CreateClient(builder,ClientUnion_Group,group1.Union());

    // Create Vector of Clients
    std::vector<flatbuffers::Offset<Client>> clients_vector;
    clients_vector.push_back(client1);
    clients_vector.push_back(client2);
    auto client_list = builder.CreateVector(clients_vector);

    auto clients = CreateClientList(builder,client_list);
    
    builder.Finish(clients);

    uint8_t *buf = builder.GetBufferPointer();
    int buf_size = builder.GetSize();

    std::ofstream fout;
    fout.open("../fb_bytes.bin", std::ios_base::binary | std::ios_base::out);

    if(fout.is_open()){
        fout.write((char*)buf, buf_size);
        fout.close();
    }else{
        std::cout<<"ERROR: not able to open bin file."<<std::endl;
    }
    return 0;
}